# gan keras로 해보기
> Not done

- toc:true
- branch: master
- badges: false
- comments: false 
- author: 최서연
- categories: [GAN]

In [1]:
import numpy as np

In [15]:
from tensorflow.keras.utils import to_categorical

In [16]:
from keras.datasets import cifar10

#### 데이터 적재

In [34]:
(x_train,y_train),(x_test,y_test)=cifar10.load_data()

데이터 셋을 적재한다.

In [35]:
x_train.shape, x_test.shape, y_train.shape,  y_test.shape

((50000, 32, 32, 3), (10000, 32, 32, 3), (50000, 1), (10000, 1))

- shape에서 첫번째 차원은 index를 가리킴.
- 두 번째 차원은 이미지 높이
- 세 번째 차원은 이미지 너비
- 마지막 차원은 RGB 채널
- 이런 배열을 4차원 tensor라 부름

In [36]:
NUM_CLASSES = 10

In [37]:
x_train = x_train.astype('float32')/255.0
x_test = x_test.astype('float32')/255.0

신경망은 -1~1 사이 놓여있게 만들기!

In [38]:
y_train = to_categorical(y_train,NUM_CLASSES)
y_test = to_categorical(y_test,NUM_CLASSES)

이미지의 정수 레이블을 one-hot-encoding vector로 바꾸기
- 어떤 이미지의 class 정수 레이블이 i라면 이 뜻은 one-hot encoding이 i번빼 원소가 1이고 그외에는 모두 0인 길이가 10(class개수)인 vector가 된다.
- 따라서 shape도 바뀐 것을 볼 수 있음

In [39]:
y_train.shape, y_test.shape

((50000, 10), (10000, 10))

#### 모델 만들기

Sequential 모델은 일렬로 층을 쌓은 네트워크를 빠르게 만들때 사용하기 좋음

API 모델은 각각 층으로 볼 수 있음

In [42]:
from tensorflow.keras.models import Sequential

In [43]:
from tensorflow.keras.layers import Flatten, Dense

In [45]:
model = Sequential([Dense(200, activation  = 'relu', input_shape=(32,32,3)),
                   Flatten(),
                   Dense(150, activation = 'relu'),
                   Dense(10,activation = 'softmax'),])

Sequential 사용한 network

In [46]:
from tensorflow.keras.layers import Input, Flatten, Dense

In [47]:
from tensorflow.keras.models import Model

In [48]:
input_layer = Input(shape=(32,32,3))

In [49]:
x=Flatten()(input_layer)

In [50]:
x=Dense(units=200,activation='relu')(x)

In [52]:
x=Dense(units=150,activation = 'relu')(x)

In [53]:
output_layer = Dense(units=10,activation='softmax')(x)

In [54]:
model=Model(input_layer,output_layer)

API 사용한 network

In [55]:
model.summary()

Model: "model"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 input_1 (InputLayer)        [(None, 32, 32, 3)]       0         
                                                                 
 flatten_1 (Flatten)         (None, 3072)              0         
                                                                 
 dense_3 (Dense)             (None, 200)               614600    
                                                                 
 dense_4 (Dense)             (None, 150)               30150     
                                                                 
 dense_5 (Dense)             (None, 10)                1510      
                                                                 
Total params: 646,260
Trainable params: 646,260
Non-trainable params: 0
_________________________________________________________________


- **Input**
    - 네트워크의 시작점
    - 입력 데이터 크기를 튜플에 알려주기
- **Flatten**
    - 입력을 하나의 vector로 펼치기
    - 32 * 32 * 3 = 3,072
- **Dense**
    - 이전 층과 완전하게 연결되는 유닛을 가짐
    - 이 층의 각 unit은 이전 층의 모든 unit과 연결됨!
    - unit의 출력은 이전 층으로부터 받은 입력과 가중치를 곱하여 더한 것
- **activation function**
    - 비선형 활성화 함수
        - 렐루Relu: 입력이 음수이면 0, 그 외는 입력과 동일한 값 출력
        - 리키렐루leaky relu: 음수이면 입력에 비례하는 작은 음수 반환, 그 외는 입력과 동일한 값 출력
        - 시그모이드sigmoid: 층의 출력을 0에서 1사이로 조정하고 시싶을 때!(binary/multilabel classification에서 사용)
        - 소프트맥스softmax: 층의 전체 출력 합이 1이 되어야 할 떄 사용(multiclass classification)

![](http://androidkt.com/wp-content/uploads/2020/05/Selection_019-1024x490.png)

![](https://kjhov195.github.io/post_img/200107/image2.png)

![](https://miro.medium.com/max/1400/1*IWUxuBpqn2VuV-7Ubr01ng.png)